### ISSUE TITLE
[TEST-FAILURE] MapStoreTest.testGetAllKeys

### ISSUE THREAD

**Issue by bilalyasar** [hazelcast/hazelcast#4024]
```
java.lang.Exception: test timed out after 120000 milliseconds
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:503)
	at com.hazelcast.spi.impl.ProxyServiceImpl$DistributedObjectFuture.get(ProxyServiceImpl.java:431)
	at com.hazelcast.spi.impl.ProxyServiceImpl$ProxyRegistry.getOrCreateProxy(ProxyServiceImpl.java:316)
	at com.hazelcast.spi.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:131)
	at com.hazelcast.instance.HazelcastInstanceImpl.getDistributedObject(HazelcastInstanceImpl.java:390)
	at com.hazelcast.instance.HazelcastInstanceImpl.getMap(HazelcastInstanceImpl.java:178)
	at com.hazelcast.instance.HazelcastInstanceProxy.getMap(HazelcastInstanceProxy.java:71)
	at com.hazelcast.map.mapstore.MapStoreTest.testGetAllKeys(MapStoreTest.java:956)
```
stacktrace + full log:
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x/com.hazelcast$hazelcast/2195/testReport/junit/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by mesutcelik** [comment]
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK8/579/com.hazelcast$hazelcast/testReport/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by ajermakovics** [comment]
According to thread dump on the last failure the loaders are still sleeping when the test finishes
```
java.lang.Thread.State: TIMED_WAITING
		at java.lang.Thread.sleep(Native Method)
		at com.hazelcast.map.mapstore.SimpleMapLoader.loadAll(SimpleMapLoader.java:31)
		at com.hazelcast.map.impl.MapStoreWrapper.loadAll(MapStoreWrapper.java:143)
```

**Comment by jerrinot** [comment]
It seems we have a regression here. The test used to be problematic in the past. It was fixed by #4082, but apparently we have a regression here. 

Here is another failure: https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-pr-builder/13832/com.hazelcast$hazelcast/testReport/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by mesutcelik** [comment]
sure.. That hasn't failed recently but started failing during last several days...

**Comment by jerrinot** [comment]
yeah, I think I found a race condition (with help of @ajermakovics) in a loading process.
I can't reproduce it yet, but I have a feasible hypothesis.

**Comment by mesutcelik** [comment]
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OracleJDK8/453/com.hazelcast$hazelcast/testReport/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by mesutcelik** [comment]
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-IbmJDK1.7/576/com.hazelcast$hazelcast/testReport/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by jerrinot** [comment]
Finally a reliable reproducer: 3d624e02b09894c33a837c4b9396026fcc0d7fed

**Comment by jerrinot** [comment]
It's caused by a possible operation reordering when a partition is migrated. 

I can see log messages like this:
```
15:39:14,628 DEBUG [LoadAllOperation] hz._hzInstance_14_dev.partition-operation.thread-49 - [127.0.0.1]:5020 [dev] [3.5-EA2-SNAPSHOT] Running LoadAllOperation for partition 49, last batch = true
15:39:14,634 DEBUG [LoadAllOperation] hz._hzInstance_14_dev.partition-operation.thread-49 - [127.0.0.1]:5020 [dev] [3.5-EA2-SNAPSHOT] Running LoadAllOperation for partition 49, last batch = false
```

Note the operation with the `last batch` set to `true` is processed before the operation where this flag is set to `false`

If I call `warmUpPartitions()` before calling getting map as in d180b9faae140c16b82d2a9fb8ee163682c5ceb5 then the issue is no longer reproducible.

**Comment by bilalyasar** [comment]
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK6/com.hazelcast$hazelcast/567/testReport/junit/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by jerrinot** [comment]
There is a PR with a fix waiting to be merged:
https://github.com/hazelcast/hazelcast/pull/5489

**Comment by bilalyasar** [comment]
@jerrinot can you please send backport of corresponding PR. 
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.maintenance-OracleJDK1.6/com.hazelcast$hazelcast/260/testReport/junit/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by jerrinot** [comment]
fixed by #5699

**Comment by hasancelik** [comment]
```
org.junit.runners.model.TestTimedOutException: test timed out after 120000 milliseconds
	at java.lang.Thread.sleep(Native Method)
	at com.hazelcast.partition.impl.InternalPartitionServiceImpl.sleepWithBusyWait(InternalPartitionServiceImpl.java:1176)
	at com.hazelcast.partition.impl.InternalPartitionServiceImpl.prepareToSafeShutdown(InternalPartitionServiceImpl.java:1122)
	at com.hazelcast.instance.Node.shutdown(Node.java:364)
	at com.hazelcast.instance.LifecycleServiceImpl.shutdown(LifecycleServiceImpl.java:91)
	at com.hazelcast.instance.HazelcastInstanceProxy.shutdown(HazelcastInstanceProxy.java:251)
	at com.hazelcast.map.mapstore.MapStoreTest.testGetAllKeys(MapStoreTest.java:504)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/718/testReport/junit/com.hazelcast.map.mapstore/MapStoreTest/testGetAllKeys/

**Comment by jerrinot** [comment]
It seems failure https://github.com/hazelcast/hazelcast/issues/4024#issuecomment-154998713 had been a side-effect of new node state transition logic and it was fixed by #6685
